src/Utilities/GenericResize.hpp

```diff
@@ -0,0 +1,88 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <array>
+#include <complex>
+#include <cstddef>
+#include <type_traits>
+
+#include "Utilities/Gsl.hpp"
+#include "Utilities/MakeWithValue.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
+/// Implementations of \link generic_resize \endlink
+///
+/// Specializations may be trivial, meaning the object is similar to a
+/// `double` and doesn't actually represent grid data.  The only
+/// requirement for a trivial specializations is defining `is_trivial`
+/// as `true`.
+///
+/// Non-trivial specializations must define `is_trivial` to `false`,
+/// and must define a static `apply` function with the signature shown
+/// in the below example.  Non-trivial specializations must also be
+/// accompanied by a specialization of the
+/// MakeWithValueImpls::NumberOfPoints struct.  The `apply` function
+/// will only be called if the current size is incorrect, so
+/// implementations should not check the current size.
+///
+/// \snippet Test_GenericResize.cpp GenericResizeImpl
+namespace GenericResizeImpls {
+/// Default implementation is not defined.
+template <typename T, typename = std::nullptr_t>
+struct GenericResizeImpl;
+}  // namespace GenericResizeImpls
+
+/// \ingroup DataStructuresGroup
+/// Change the number of grid points in an object.
+///
+/// Change the number of points stored in \p result to a given value
+/// or to match another object.  If \p pattern is a `size_t` it will
+/// be used as the number of points, otherwise it will be interpreted
+/// as data on a collection of grid points.
+///
+/// \see GenericResizeImpls, make_with_value
+template <typename T, typename U>
+void generic_resize(const gsl::not_null<T*> result, const U& pattern) {
+  if constexpr (not GenericResizeImpls::GenericResizeImpl<T>::is_trivial) {
+    const size_t size = MakeWithValueImpls::number_of_points(pattern);
+    if (UNLIKELY(MakeWithValueImpls::number_of_points(*result) != size)) {
+      GenericResizeImpls::GenericResizeImpl<T>::apply(result, size);
+    }
+  }
+}
+
+template <>
+struct GenericResizeImpls::GenericResizeImpl<double> {
```

Comment:
Maybe also add a specialization for `std::vector`? I think that, along with the implementations in the next commit, covers all the common containers we'd normally use.